File: typescript/leetcode/lib/Tree.ts
export interface TreeNode<T> {
  val: T
  left: TreeNode<T> | null
  right: TreeNode<T> | null
}

export function createTree<T>(
  values: T[],
  root: TreeNode<T> | null = null,
  index: number = 0
): TreeNode<T> | null {
  if (index < values.length) {
    root = { val: values[index], left: null, right: null }
    root.left = createTree(values, root.left, 2 * index + 1)
    root.right = createTree(values, root.right, 2 * index + 2)
  }

  return root
}

export function toArray<T>(
  tree: TreeNode<T> | null,
  values: T[] = [],
  index: number = 0
): T[] {
  if (tree != null) {
    values[index] = tree.val
    toArray(tree.left, values, 2 * index + 1)
    toArray(tree.right, values, 2 * index + 2)
  }

  return values
}

File: python/leetcode/lib/tree.py
from typing import Generic, TypeVar, Optional, List

T = TypeVar("T")


class TreeNode(Generic[T]):
    val: Optional[T]
    left: Optional["TreeNode[T]"]
    right: Optional["TreeNode[T]"]

    def __init__(self, val: T):
        self.val = val
        self.left = None
        self.right = None


root = TreeNode(10)
root.left


def create_tree(values: List[T], index: int = 0) -> Optional[TreeNode[T]]:
    if len(values) - 1 < index:
        return None

    root = TreeNode(values[index])
    root.left = create_tree(values, index * 2 + 1)
    root.right = create_tree(values, index * 2 + 2)

    return root


def create_list(root, values=None, index=0) -> list:
    if not values:
        values = []

    if not root:
        return values

    size = len(values) - 1
    if size <= index:
        values += [None] * (index - size)

    values[index] = root.val
    create_list(root.left, values, index * 2 + 1)
    create_list(root.right, values, index * 2 + 2)

    return values
